### PR TITLE
Plush Ratvar can now fight plush Nar-Sie

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: Ratvar and Nar-Sie plushes now have a unique interaction.
/:cl:

I thought that this would be amusing, so I added it. If a plushie of Ratvar encounters a plushie of Nar-Sie, they will clash like their namesakes! It has the same sounds, but they only play locally and play at double speed (so they sound really high-pitched.) One plushie hitting another is represented by the hittee doing a spin.

Upon winning, the losing plushie will explode. Victory chances start at 10% and increase by 10% every exchange of attacks; there's a max of 10 hits between the two before a victor is just chosen randomly.

### CODE REVIEW

**Comment by deathride58** [comment]
Speedmerge this or riot

**Comment by RogueSteampunker** [comment]
"Alright crew, we're divided on leaving or staying. As such, we must make the plushies fight to decide.

May the best god win."

**Comment by DoctoreMad** [comment]
If you made the RatVar/NarSie cultist amount affect the plushie win rates, would it be metagamed only a bit or a lot?

**Comment by nicbn** [comment]
A lot

**Comment by Alek2ander** [comment]
I'm detecting inexcusable levels of bias here.
Ratvar has almost 60% winrate.

Seriously, because Ratvar always does his rolls first, he has a pretty big advantage here.
This is **important**

**Comment by Doppeltongue** [comment]
Make them roll for initiative?

**Comment by Shadowlight213** [comment]
Oh my god thos is adorable. Speedmerge post haste

**Comment by Ratvar** [comment]
Can you https://github.com/tgstation/tgstation/pull/31986 activate grenade of lost plushie?

**Comment by Bawhoppen** [comment]
:confetti_ball:

**Comment by KorPhaeron** [comment]
The chance of a maintainer willing to review 62 lines of stuffed animals processing/fighting seem incredibly low

**Comment by kilkun** [comment]
ratvar sux nar-sie is way cooler make him have the +10% winrate pls

File: code/game/objects/items/plushes.dm [hunk 1]
```diff
@@ -35,6 +35,68 @@
...
+
+/obj/item/toy/plush/plushvar/Initialize()
+	. = ..()
+	START_PROCESSING(SSobj, src)
+
+/obj/item/toy/plush/plushvar/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	. = ..()
+
+/obj/item/toy/plush/plushvar/process()
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: Just do this in Moved instead and also on the nar plush

File: code/game/objects/items/plushes.dm [hunk 2]
```diff
@@ -35,6 +35,68 @@
...
+	var/victory_chance = 10
+	for(var/i in 1 to 10) //We only fight ten times max
+		if(!src || !P)
+			return
+		if(!Adjacent(P))
+			visible_message("<span class='warning'>The two plushies angrily flail at each other before giving up.</span>")
+			clash_target = null
+			return
+		playsound(src, 'sound/magic/clockwork/ratvar_attack.ogg', 50, TRUE, frequency = 2)
+		sleep(2.4)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: timers instead of sleeps

File: code/game/objects/items/plushes.dm [hunk 3]
```diff
@@ -35,6 +35,68 @@
...
+		if(prob(victory_chance))
+			a_winnar_is = src
+			break
+		P.SpinAnimation(5, 0)
+		sleep(5)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: ditto

File: code/game/objects/items/plushes.dm [hunk 4]
```diff
@@ -35,6 +35,68 @@
...
+		playsound(P, 'sound/magic/clockwork/narsie_attack.ogg', 50, TRUE, frequency = 2)
+		sleep(3.3)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: ditto

File: code/game/objects/items/plushes.dm [hunk 5]
```diff
@@ -35,6 +35,68 @@
...
+		if(prob(victory_chance))
+			a_winnar_is = P
+			break
+		SpinAnimation(5, 0)
+		victory_chance += 10
+		sleep(5)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: ditto

File: code/game/objects/items/plushes.dm [hunk 6]
```diff
@@ -35,6 +35,68 @@
...
+		playsound(P, 'sound/magic/demon_dies.ogg', 50, TRUE, frequency = 2)
+		explosion(get_turf(P), 0, 0, 1)
+		qdel(P)
+		clash_target = null
+	else
+		say("NO! I will not be banished again...")
+		P.say(pick("Ha.", "Ra'sha fonn dest.", "You fool. To come here."))
+		playsound(src, 'sound/magic/clockwork/anima_fragment_death.ogg', 50, TRUE, frequency = 2)
+		playsound(P, 'sound/magic/demon_attack1.ogg', 50, TRUE, frequency = 2)
+		explosion(get_turf(src), 0, 0, 1)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: explosion(src, ...

File: code/game/objects/items/plushes.dm [hunk 7]
```diff
@@ -35,6 +35,68 @@
...
+		victory_chance += 10
+		sleep(5)
+	if(!a_winnar_is)
+		a_winnar_is = pick(src, P)
+	if(a_winnar_is == src)
+		say(pick("DIE.", "ROT."))
+		P.say(pick("Nooooo...", "Not die. To y-", "Die. Ratv-", "Sas tyen re-"))
+		playsound(src, 'sound/magic/clockwork/anima_fragment_attack.ogg', 50, TRUE, frequency = 2)
+		playsound(P, 'sound/magic/demon_dies.ogg', 50, TRUE, frequency = 2)
+		explosion(get_turf(P), 0, 0, 1)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: explosion(P, ...

File: code/game/objects/items/plushes.dm [hunk 8]
```diff
@@ -35,6 +35,68 @@
...
 	icon_state = "plushvar"
+	var/obj/item/toy/plush/narplush/clash_target
+
+/obj/item/toy/plush/plushvar/Initialize()
+	. = ..()
+	START_PROCESSING(SSobj, src)
+
+/obj/item/toy/plush/plushvar/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	. = ..()
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: return ..()

File: code/game/objects/items/plushes.dm [hunk 9]
```diff
@@ -35,6 +35,68 @@
...
 	desc = "An adorable plushie of the clockwork justiciar himself with new and improved spring arm action."
 	icon_state = "plushvar"
+	var/obj/item/toy/plush/narplush/clash_target
+
+/obj/item/toy/plush/plushvar/Initialize()
+	. = ..()
+	START_PROCESSING(SSobj, src)
+
+/obj/item/toy/plush/plushvar/Destroy()
+	STOP_PROCESSING(SSobj, src)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: needs to null `clash_target`

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/objects/items/plushes.dm (hunk 2 above)
Comment (Xhuis) [reply code/game/objects/items/plushes.dm]: That's a great deal of timers, and there'd need to be a string of different procs to do it. Would `set waitfor = FALSE` not be a cleaner solution to do the same thing?

**Comment by Cyberboss** [comment]
in response to your outdated review comment, timers are always more efficient than sleeps

**Comment by Hatterhat** [comment]
merge when

**Comment by Xhuis** [comment]
I'll see about timers shortly.

**Comment by JJRcop** [comment]
@Xhuis You can always make timers that call the same proc but give a different argument for the stage

File: code/game/objects/items/plushes.dm (hunk 2 above)
Comment (JJRcop) [reply code/game/objects/items/plushes.dm]: BTW the reason he wants timers instead of set waitfor = FALSE is because of an unrelated project

**Comment by Xhuis** [comment]
I might just be missing something obvious here, but wouldn't timers make it so I have to cache victory chances etc. on the plushie itself, instead of just in the proc? Is that still worth using them?

**Comment by JJRcop** [comment]
Xhuis, you give the victory chance as an argument in the timer

**Comment by JJRcop** [comment]
In fact 3 extra args, you need to pass a stage, ```i```, and the victory chance. That is if you do decide to go for the 1 proc multiple stages thing

**Comment by Xhuis** [comment]
Hmmm.

**Comment by optimumtact** [comment]
timers are absoluteyl not always more efficient than sleeps and in this case they'd make the code hideous.

We could benefit from a proper actors framework for scripting scenes.

side note, you need to make sure both objects are not !QDELETED after every sleep

**Pull request merged**